src/ui/main.qml

```diff
@@ -65,6 +66,23 @@ Window {
         if (VPN.startMinimized)
             this.showMinimized();
 
+        Glean.initialize(VPN.versionString, {
+            post(url, body, headers) {
+                return new Promise((resolve, reject) => {
+                    const xhr = new XMLHttpRequest();
+                    xhr.open("POST", url);
+
+                    for (const header in headers) {
+                      xhr.setRequestHeader(header, headers[header]);
+                    }
+                    xhr.onloadend = () => {
```

Comment:
Is `onloadend` fired when there is a timeout or when network is not available? Note that, this function should always resolve and return _some_ status to Glean.

Reply:
onloadend is called when the XHR completes the task, with a failure or with a success.

Reply:
Yes, but timeout [means there the task was never completed](https://xhr.spec.whatwg.org/#events).

Reply:
Does glean.js check the status code to be == 20x?

Reply:
> 
> 
> Does glean.js check the status code to be == 20x?

Glean's upload logic needs to be aware of the status code because it will make decisions based on that, especially whether or not to retry upload or to remove the stored ping. However, our Qt platform implementation has no storage so this should not make too much of a difference.

I believe the uploader implementation should be provided by Glean.js (we do that for other platforms, too). For the sake of unblocking you, I believe it's fine for it to be as it is. If you want a reference, here's the current [webext uploader](https://github.com/mozilla/glean.js/blob/main/glean/src/platform/webext/uploader.ts).

Reply:
I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1709371